File: leetcode/golang/82_remove_duplicates_from_sorted_list_ii.go

type ListNode struct {
    Val int
    Next *ListNode
}

func deleteDuplicates(head *ListNode) *ListNode {
    node := &ListNode {
        Val : 0,
        Next : head,
    }
    cur := node
    for head != nil {
        if head.Next != nil && (head.Next.Val == head.Val) {
            x := head.Val
            for head != nil && head.Val == x {
                head = head.Next
            }
            cur.Next = head
        } else {
            cur = cur.Next
            head = head.Next
        }
    }
    return node.Next
}

File: leetcode/c++/82_remove_duplicates_from_sorted_list_ii.cpp
// https://leetcode.cn/problems/remove-duplicates-from-sorted-list-ii/

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == nullptr) return head;
        
        ListNode* dummy = new ListNode(0, head); 
        ListNode* cur = dummy; 
        while (head) {
            if (head->next && head->val == head->next->val) {
                int x = head->val;
                while (head && head->val == x) { 
                    head = head->next;
                }
                cur->next = head; 
            } else {
                cur = cur->next; 
                head = head->next; 
            }
        }
        
        return dummy->next;
    }
};

// O(n), O(1)
